provisioner/terraform/executor.go

```diff
@@ -334,6 +334,68 @@ func onlyDataResources(sm tfjson.StateModule) tfjson.StateModule {
 	return filtered
 }
 
+func (e *executor) logResourceReplacements(ctx context.Context, plan *tfjson.Plan) {
+	if plan == nil {
+		return
+	}
+
+	if len(plan.ResourceChanges) == 0 {
+		return
+	}
+	var (
+		count        int
+		replacements = make(map[string][]string, len(plan.ResourceChanges))
+	)
+
+	for _, ch := range plan.ResourceChanges {
+		// No change, no problem!
+		if ch.Change == nil {
+			continue
+		}
+
+		// No-op change, no problem!
+		if ch.Change.Actions.NoOp() {
+			continue
+		}
+
+		// No replacements, no problem!
+		if len(ch.Change.ReplacePaths) == 0 {
+			continue
+		}
+
+		// Replacing our resources, no problem!
+		if strings.Index(ch.Type, "coder_") == 0 {
+			continue
+		}
+
+		for _, p := range ch.Change.ReplacePaths {
+			var path string
+			switch p := p.(type) {
+			case []interface{}:
+				segs := p
+				list := make([]string, 0, len(segs))
+				for _, s := range segs {
+					list = append(list, fmt.Sprintf("%v", s))
+				}
+				path = strings.Join(list, ".")
+			default:
+				path = fmt.Sprintf("%v", p)
+			}
+
+			replacements[ch.Address] = append(replacements[ch.Address], path)
+		}
+
+		count++
+	}
+
+	if count > 0 {
+		e.server.logger.Warn(ctx, "plan introduces resource changes", slog.F("count", count))
```

Comment:
why are we dropping these logs at Warn level? This is just for debugging, right? Like, it's not unusual for a plan to change the resources around.

Reply:
In this case, a replacement is a pathological scenario with pre-provisioned resources now drifting and needing to be replaced. See https://coder.com/docs/@main/admin/templates/extending-templates/prebuilt-workspaces#preventing-resource-replacement.

https://github.com/coder/coder/pull/17571 expands upon this by adding a metric & an alert to make things more visible, but it's worth logging a warning as well.

Reply:
Yeah, but as far as I can tell, this happens for _every_ build, not just claim-a-prebuilt-workspace builds.

Reply:
If https://github.com/coder/coder/pull/17571 lands in time, we can remove this logging functionality from this branch since my PR is a more full-featured implementation, and only logs these changes on prebuild claim runs.
cc @SasSwart

Reply:
Bringing this back to your attention. Is this needed in this PR in light of #17571 ?

Reply:
We'll leave it in here for now and @dannykopping will update when he merges https://github.com/coder/coder/pull/17571.

Reply:
https://github.com/coder/coder/pull/17571/commits/b9eb8bee69fda550ea41c466d684641ccbd9ab3c